python/singa/autograd.py

```diff
@@ -743,25 +755,43 @@ def __call__(self, x):
         self.device_check(x, self.W, self.b)
 
         if x.device.id() == -1:
-            if not hasattr(self, 'handle'):
-                self.handle = singa.ConvHandle(x.data, self.kernel_size, self.stride,
-                                               self.padding, self.in_channels, self.out_channels, self.bias)
-            elif x.shape[0] != self.handle.batchsize:
-                self.handle = singa.ConvHandle(x.data, self.kernel_size, self.stride,
-                                               self.padding, self.in_channels, self.out_channels, self.bias)
+            if self.groups != 1:
+                raise ValueError('Not implemented yet')
+            else:
+                if not hasattr(self, 'handle'):
+                    self.handle = singa.ConvHandle(x.data, self.kernel_size, self.stride,
+                                                   self.padding, self.in_channels, self.out_channels, self.bias)
+                elif x.shape[0] != self.handle.batchsize:
+                    self.handle = singa.ConvHandle(x.data, self.kernel_size, self.stride,
+                                                   self.padding, self.in_channels, self.out_channels, self.bias)
         else:
             if not hasattr(self, 'handle'):
                 self.handle = singa.CudnnConvHandle(x.data, self.kernel_size, self.stride,
-                                                    self.padding, self.in_channels, self.out_channels, self.bias)
+                                                    self.padding, self.in_channels, self.out_channels, self.bias, self.groups)
             elif x.shape[0] != self.handle.batchsize:
                 self.handle = singa.CudnnConvHandle(x.data, self.kernel_size, self.stride,
-                                                    self.padding, self.in_channels, self.out_channels, self.bias)
+                                                    self.padding, self.in_channels, self.out_channels, self.bias, self.groups)
         self.handle.device_id = x.device.id()
 
         y = conv2d(self.handle, x, self.W, self.b)
         return y
 
 
+class SeparableConv2d(Layer):
+
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, bias=False):
+
+        self.mapping_spacial_conv = Conv2d(
```

Comment:
rename mapping_spacial_conv to spatial_conv?

Reply:
ok.